src/main/java/com/vmware/vipclient/i18n/messages/api/opt/server/RemotePatternOpt.java

```diff
@@ -4,82 +4,88 @@
+                responseCode.equals(HttpURLConnection.HTTP_NOT_MODIFIED))) {
+            long timestamp = 0;
+            String etag = null;
+            Long maxAgeMillis = null;
+
+            if (response.get(URLUtils.RESPONSE_TIMESTAMP) != null)
+                timestamp = (long) response.get(URLUtils.RESPONSE_TIMESTAMP);
+            if (response.get(URLUtils.HEADERS) != null)
+                etag = URLUtils.createEtagString((Map<String, List<String>>) response.get(URLUtils.HEADERS));
+            if (response.get(URLUtils.MAX_AGE_MILLIS) != null)
+                maxAgeMillis = (Long) response.get(URLUtils.MAX_AGE_MILLIS);
+
+            if (responseCode.equals(HttpURLConnection.HTTP_OK)) {
+                try {
+                    String responseBody = (String) response.get(URLUtils.BODY);
+                    Map<String, Object> patterns = getPatternsFromResponse(responseBody);
+                    if (patterns != null) {
+                        logger.debug("Find the pattern from Singleton Service for locale [{}].\n", locale);
+                        cacheItem.set(patterns, etag, timestamp, maxAgeMillis);
+                    }else{
+                        logger.debug("Doesn't find the pattern from Singleton Service for locale [{}].\n", locale);
+                        cacheItem.set(etag, timestamp, maxAgeMillis);
+                    }
+                } catch (Exception e) {
+                    logger.error("Failed to get pattern data from remote!");
+                }
+            }else{
+                logger.debug("There is no update on Singleton Service for the pattern of locale [{}].\n", locale);
+                cacheItem.set(etag, timestamp, maxAgeMillis);
             }
-            return msgObject;
         }
     }
 
-    private Object getCategoriesFromResponse(String responseStr, String node) {
-        Object msgObject = null;
-        try {
-            JSONObject responseObj = (JSONObject) JSONValue.parseWithException(responseStr);
-            if (responseObj != null) {
-                Object dataObj = responseObj.get(ConstantsKeys.DATA);
-                if (dataObj != null && dataObj instanceof JSONObject) {
-                    msgObject = ((JSONObject) dataObj).get(node);
-                }
+    private Map<String, Object> getPatternsFromResponse(String responseBody) {
+        Map<String, Object> categoriesObj = null;
+        Map<String, Object> dataObj = (Map<String, Object>) getDataFromResponse(responseBody);
+        if (dataObj != null && dataObj instanceof JSONObject) {
+            Object obj = dataObj.get(PatternKeys.CATEGORIES);
+            if (obj != null && obj instanceof JSONObject) {
+                categoriesObj = (Map<String, Object>) obj;
             }
-        } catch (ParseException e) {
-            // TODO Auto-generated catch block
-            logger.error(e.getMessage());
         }
-        return msgObject;
+        return categoriesObj;
```

Comment:
May you call BaseOpt.getMessagesFromResponse to get the data?

Reply:
@Xiaochao8 I think the method name 'getMessagesFromResponse' is not appropriate, can't represent what the method do, if I call it I'm afraid of causing confusion when read it again after some time. If I change the method's name now, will introduce L3 code change which may let the PR scope more big, so I write new method.

I also agree with @jessiejuachon we shouldn't put specific parse for one API to BaseOpt and revisit and clean this up in future.